### ISSUE TITLE
[code documentation] difference between Ident.tbl and Map.Make(Ident).t?

### ISSUE THREAD

**Issue by gasche** [ocaml/ocaml#10843]
What is the difference between `Ident.tbl` and `Map.Make(Ident).t`? Could it be documented in indent.mli or any other place? When should one use one or the other? (cc @garrigue)

From a distance, it looks like Ident.tbl is a manually-inlined definition of balanced maps, designed such that `'a Ident.tbl` is equivalent to `(Ident.t * 'a) list Map.Make(String).t`, with a slightly more compact data representation, where the string key is the "name" (the user-visible part) of the identifier. This enables using the same map to lookup either full identifiers, or just names (more recent bindings shadow older bindings). `Ident.tbl` also lacks many functions on maps, including comparison/equality functions.

From this distance still, it looks like when we are working on the typedtree (we always have full identifiers at hand), and not in the process of typing a parsetree, the two structures should be roughly equivalent: `Ident.tbl` will store shadowing bindings in a list, while `Map.Make(Ident)` will inline them in the tree structure; the only differences in storage size should be constant factors.
`Ident.tbl` is probably more memory-efficient in presence of shadowing, but there might be trouble if a *lot* of shadowing happens on a given binding, due to linear (non-tail-recursive) traversals, but this would never happen in practice, right?

Yet: in commit https://github.com/ocaml/ocaml/commit/cbc66ffdd14958120663e740e5c81ad4122df436, @yallop switched from `Ident.tbl` to `Map.Make(Ident).t`, after explaining in https://github.com/ocaml/ocaml/pull/556#issuecomment-330262556 that using `Ident.tbl` for his use-case would result in Stack Overflow on some OCaml packages (cil and frama-c-base, I guess that the latter contains a modified copy of cil). Were the stack overflows coming from the linear same-name lists, or somewhere else?

(Is there a measurable performance benefit to using this code-duplicating `tbl` definition, instead of using `(Ident.t * 'a) list String.Map.t` directly? If I count correctly, the common case of storing a value for a single identifier of a given name (no shadowing) takes 9 heap words with `tbl` (3 `'a data` fields plus their block header), and would take 12 heap words with `(Ident.t * 'a) list`, so we are duplicating code for a 25% reduction in memory usage for an in-memory data structure.)

**Comment by xavierleroy** [comment]
There are two reasons for using  Ident.tbl instead of Map.Make(Ident).t, a good one and an historical one.

The good reason is that it provides the ability to search by name as well as by name + stamp; the former is essential to implement the Env module.

The historical reason is that a long time ago this code was bootstrapped from Caml Light, which didn't have functors, so there was no generic Map and Set abstractions available.

If none of these reasons apply to your use case, it should not really matter which abstraction is used.  The difference in memory usage is probably lost in the noise.

I'm a bit surprised that Ident.tbl can overflow the stack while Map.Make(Ident).t doesn't, but I didn't look into scenarios that could cause this.

**Comment by gasche** [comment]
Thanks for the clarification.

I would be curious if @yallop could shed light on the stack overflows.

What I'm planning to do for now:
1. Send documentation comment for the `tbl` interface and implementation
    (Low-hanging fruit.)
2. Consider sending a PR to replace the bespoke implementation with a functor instantiation, and then kindly ask @Octachron to look at any performance impact. (This one is likely to be a harder sell.) I had to duplicate yet more code from `Map` to implement `Ident.compare_tbl` in #10825, and I would rather have future users spend their effort elsewhere.

**Comment by Drup** [comment]
I considered replacing the bespoke implementation several times over: it is not that easy. The dual nature of the lookup (by id and by name, including when name-shadowing occurs) makes it a bit inconvenient to encode into regular `Map`s.

I generally agree the duplication is annoying, though.

**Comment by gasche** [comment]
What's wrong with `(Ident.t * 'a) list Map.Make(String).t`?

Note: I thought of opening PRs to (1) document and (2) remove the duplication and also (3) to add `Ident.compare_tbl : ('a -> 'a -> bool) -> 'a tbl -> 'a tbl -> bool` which I needed in #10825. But I haven't done any of this yet:
- it's possible but not obvious to justify that the "if it ain't broke, don't fix it" wisdom should be ignored here
- we are in a sequential-freeze period where it is unclear those changes are welcome

I may revisit once the multicore-PR has landed.

**Comment by yallop** [comment]
@gasche:
> Yet: in commit [cbc66ff](https://github.com/ocaml/ocaml/commit/cbc66ffdd14958120663e740e5c81ad4122df436), @yallop switched from `Ident.tbl` to `Map.Make(Ident).t`, after explaining in [#556 (comment)](https://github.com/ocaml/ocaml/pull/556#issuecomment-330262556) that using `Ident.tbl` for his use-case would result in Stack Overflow on some OCaml packages (cil and frama-c-base, I guess that the latter contains a modified copy of cil). Were the stack overflows coming from the linear same-name lists, or somewhere else?

I reproduced this issue just now using 4.06.0 with cbc66ffdd14958120663e740e5c81ad4122df436 reverted.  In this case, the stack overflow arose from [`Ident.fold_data`](https://github.com/ocaml/ocaml/blob/5fae5fd90e15e956569dbbeb12ccb78b83aceeef/typing/ident.ml#L298-L301), which is not tail-recursive:

```ocaml
let rec fold_data f d accu =
  match d with
    None -> accu
  | Some k -> f k.ident k.data (fold_data f k.previous accu)
```

**Comment by gasche** [comment]
Do you understand how the stack-overflow occurs in practice? (Does it require a typing environment within a recursive value definition that has dozens of thousands of bindings?) Here is what I understand so far:

1. `fold_data` / `fold_all` are basically `Map.fold_right` for this structure (they include traversing shadowed binding).
2. But your `Map.Make(Ident)` version also includes shadowed bindings in the map, so using `fold_right` there should have roughly similar stack consumption. If that sufficed to explain the solution, there you are only constant-factor better and we are very lucky that the new version does not blow up; I would guess that something else is going on.
3. The PR also change the rest of the code to `fold` less often; in particular you used it to implement `map` (which is also not provided by `Ident.tbl`, forcing you to reinvent it less efficiently), and the new code uses `Map.map` which is logarithmic in stack space.
4. but the new code still uses `fold` to implement `join` (union), which is also stack-space linear in the worst case (in its *left* parameter), so we could see blowups there.

This is mysterious to me.

A wild guess: in your original work on recursive value checking (before we collaborated on a simpler way to do things), the types / summaries inferred for expressions where, if I remember correctly, maps of maps. Is it possible that in certain cases we would fold over a fold, and that this would somehow result in stack-space usage that is *quadratic* over the map size? (I'm looking at [this code](https://github.com/ocaml/ocaml/blob/cbc66ffdd14958120663e740e5c81ad4122df436/typing/typecore.ml#L1807-L1816) for example.)

This would plausibly explain a stack overflow on large human-written inputs, but: as far as I can tell, if `f` has worse-case stack-space complexity `O(f)`, and `fold` has worst-case stack-stack complexity `O(g)`, then I would expect that the stack consumption of `fold f` is `O(f + g)`, not `O(f * g)` -- the recursive call with shape `f xs (fold f x accu)` calls `fold` first, then it can reclaim stack space, then calls `f`.

**Comment by yallop** [comment]
I haven't investigated closely, but it may simply come down to the fact that `Ident.add` extends the table with a new binding (keeping previous bindings for the same name around), while `Map.Make(Ident).add` replaces the binding.  Hence the following code

```ocaml
module M = Map.Make(Ident)
module I = Ident

let x = I.create "x"
let xs = List.init 100000 (fun _ -> x)
let i = List.fold_left (fun t x -> I.add x () t) I.empty xs
let m = List.fold_left (fun t x -> M.add x () t) M.empty xs
let () = Printf.printf "size(i): %d\nsize(m): %d\n"
           Obj.(reachable_words (repr i)) Obj.(reachable_words (repr m))

```

prints

```
size(i): 600009
size(m): 12
```

For the let-rec check, calling `add` (in `join`) for names that are already present in the map leads `Ident` to build large data structures that keep the full history of the mode inference for each identifier.

**Comment by gasche** [comment]
That makes sense, thanks!

If I understand correctly, the problem comes from using an `add` function that does not properly replace existing bindings (of the same identifier with the same stamp) to implement union. It could have been fixed by the following diff

```diff
    let join x y =
      Ident.fold_all
        (fun id v tbl ->
           let v' = try Ident.find_same id tbl with Not_found -> Guarded in
-           Ident.add id (prec v v') tbl)
+           Ident.add id (prec v v') (Ident.remove_same id tbl))
        x y
```

except that the function `Ident.remove_same` does not exist.

**Comment by xavierleroy** [comment]
>  Ident.add extends the table with a new binding (keeping previous bindings for the same name around), while Map.Make(Ident).add replaces the binding. 

Yes, that's the reason.  `Ident.add` doesn't check for previous bindings of the same name+stamp identifier because, in the way Ident was originally used to implement Env, multiple bindings of the same identifier never occur, since identifiers are unique...  Also, it may not be algorithmically efficient (linear in the number of identifiers with the same name).

> except that the function Ident.remove_same does not exist.

Again, it would not have good complexity...

The bottom line is that a regular `Map.Make(Ident)` is to be preferred when the extra capabilities of `Ident` are not needed, namely lookup by name.

**Comment by gasche** [comment]
I agree, but the details of the discussion show that it is important to document this behavior of `Ident.add` so that users can understand why their code may blow up.

> Again, [remove_same] would not have good complexity...

This is the same complexity as `find_same`, and it is explicit, so it would be fine. (But I did not propose to make `add` replace same-bindings by default, which could add a constant factor overhead to many situations where the current approach is fine.)

**Comment by xavierleroy** [comment]
> This is the same complexity as find_same, and it is explicit, so it would be fine.

Point well taken.  I forgot that `find_same` has bad worst-case too...

**Comment by gasche** [comment]
I would expect that shadowing is rare enough in practice that those lists remain small. (The worst case would be a computer-generated program that repeats `let state = <some udpate function> state in`, and even that would hit other issues before hitting the stack-space limit on shadowing.)

On the other hand, with Jeremy's non-replacing implementation of `join` used in program analysis, each binding would effectively get duplicated by the number of nodes on the AST, leading to a quadratic blowup in total map size -- and then a stack overflow on folding.

**Comment by gasche** [comment]
I proposed some documentation in #11179, which is now merged. (Outside that PR) I also wrote an implementation of `remove_same` along the lines discussed here, but I was too lazy to test it (there are no unit tests for `Ident` in the testsuite yet) so I decided to wait until there is demand.
